### PR TITLE
*: change the interface for kv.Response

### PR DESCRIPTION
**Pull request by lamxTyler**
`Next` for the `kv.Response` returns an interface for result subset in this PR.
PTAL @shenli @coocood

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -466,10 +466,17 @@ func (it *copIterator) sendToRespCh(resp copResponse, respCh chan copResponse) (
 	return
 }
 
+type copResultSubset struct {
+	data []byte
+}
+
+func (rs copResultSubset) GetData() []byte {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: rs *copResultSubset

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -516,9 +523,9 @@ func (it *copIterator) Next(ctx context.Context) ([]byte, error) {
 	}
 
 	if resp.Data == nil {
-		return []byte{}, nil
+		return copResultSubset{}, nil
```
Comment (shenli) [inline store/tikv/coprocessor.go]: return pointer.

File: kv/kv.go [hunk 3]
```diff
@@ -200,12 +200,18 @@ type Request struct {
 	Streaming bool
 }
 
+// ResultSubset represents a result subset from a single storage unit.
+// TODO: Find a better interface for ResultSubset that can avoid allocation and reuse bytes.
```
Comment (shenli) [inline kv/kv.go]: There is no allocation for the byte slice.

File: distsql/distsql.go [hunk 4]
```diff
@@ -150,7 +150,10 @@ func (r *selectResult) NextRaw(ctx context.Context) ([]byte, error) {
 	re := <-r.results
 	r.partialCount++
 	r.scanKeys = -1
-	return re.result, errors.Trace(re.err)
+	if re.result == nil || re.err != nil {
+		return nil, errors.Trace(re.err)
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply distsql/distsql.go]: Trace nil error is still nil.

**Comment by shenli** [comment]
@coocood PTAL

File: store/tikv/coprocessor.go [hunk 5]
```diff
@@ -466,10 +466,17 @@ func (it *copIterator) sendToRespCh(resp copResponse, respCh chan copResponse) (
 	return
 }
 
+type copResultSubset struct {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Add comments for this structure like "copResultSubset implements `kv.ResultSubset` interface."

File: distsql/distsql.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhexuany) [reply distsql/distsql.go]: That is not my point. If result being nil is legitimate, then that is fine. I am worried about result being nil could be problematic and we did not trace this error.

[the remark this answers is not in the document]
Comment (lamxTyler) [reply distsql/distsql.go]: Result is nil is legitimate.

**Comment by coocood** [comment]
LGTM

**Pull request merged**